cpp/src/reductions/scan/scan_inclusive.cu

```diff
@@ -330,6 +330,270 @@ std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                                nullcnt.end() - 1,
+                                nullcnt.begin(),
+                                nullcnt.begin() + 1);
+
+  valid_it = cudf::detail::make_validity_iterator(device_view);
+  thrust::transform(
+    rmm::exec_policy(stream),
+    valid_it,
+    valid_it + input.size(),
+    pairs.begin(),
+    pairs.begin(),
+    [=] __host__ __device__(bool valid,
+                            thrust::pair<double, double> pair) -> thrust::pair<double, double> {
+      if (!valid) {
+        return thrust::pair<double, double>(1.0, 0.0);
+      } else {
+        return pair;
+      }
+    });
+
+  valid_it           = cudf::detail::make_validity_iterator(device_view);
+  auto valid_and_exp = thrust::make_zip_iterator(thrust::make_tuple(valid_it, nullcnt.begin()));
+
+  thrust::transform(
+    rmm::exec_policy(stream),
+    valid_and_exp,
+    valid_and_exp + input.size(),
+    pairs.begin(),
+    pairs.begin(),
+    [=] __host__ __device__(thrust::tuple<bool, int> valid_and_exp,
+                            thrust::pair<double, double> pair) -> thrust::pair<double, double> {
+      bool valid = thrust::get<0>(valid_and_exp);
+      int exp    = thrust::get<1>(valid_and_exp);
+      if (valid & (exp != 0)) {
+        double beta  = thrust::get<0>(pair);
+        double scale = thrust::get<1>(pair);
+        return thrust::pair<double, double>(beta * (pow(beta, exp)), scale);
+      } else {
+        return pair;
+      }
+    });
+}
+
+void compute_ewma_adjust(column_view const& input,
+                         double beta,
+                         mutable_column_view output,
+                         rmm::cuda_stream_view stream,
+                         rmm::mr::device_memory_resource* mr)
+{
+  rmm::device_uvector<thrust::pair<double, double>> pairs(input.size(), stream, mr);
+
+  // Numerator
+  // Fill with pairs
+  thrust::transform(rmm::exec_policy(stream),
+                    input.begin<double>(),
+                    input.end<double>(),
+                    pairs.begin(),
+                    [=] __host__ __device__(double input) -> thrust::pair<double, double> {
+                      return thrust::pair<double, double>(beta, input);
+                    });
+
+  if (input.has_nulls()) { pair_beta_adjust(input, pairs, beta, stream); }
+
+  compute_recurrence(pairs, stream);
+
+  // copy the second elements to the output for now
+  thrust::transform(rmm::exec_policy(stream),
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin<double>(),
+                    [=] __host__ __device__(thrust::pair<double, double> pair) -> double {
+                      return thrust::get<1>(pair);
+                    });
+
+  // Denominator
+  // Fill with pairs
+  thrust::fill(
+    rmm::exec_policy(stream), pairs.begin(), pairs.end(), thrust::pair<double, double>(beta, 1.0));
+
+  if (input.has_nulls()) { pair_beta_adjust(input, pairs, beta, stream); }
```

Comment:
Filling the pair vector with a constant value seems wasteful. If `pair_beta_adjust` returned its output and took an iterator as input instead of a vector, you could pass a `constant_iterator` as input to this call to `pair_beta_adjust` and avoid the fill.  (In general when I see calls to `thrust::fill` or `thrust::sequence` I look for more efficient approaches, often involving iterators)